src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,275 @@
+        let metadata = TreeMeta::new(root_offset, self.leaves_offset, tree_offset);
+        let meta_buf = serialize(&metadata)?;
+        Ok(TreeStage {
+            headers_iter: self.headers_iter,
+            metadata,
+            meta_buf,
+            tree_buf,
+            header: self.header,
+            headers_size: self.headers_size,
+            leaves_buf: self.leaves_buf,
+            meta: self.meta,
+        })
+    }
+
+    fn serialize_bptree(
+        keys: Vec<K>,
+        leaves_offset: u64,
+        leaf_size: u64,
+        tree_offset: u64,
+    ) -> Result<(u64, Vec<u8>)> {
+        let max_amount = Self::max_leaf_node_capacity(keys[0].as_ref().len());
+        let nodes_amount = (keys.len() - 1) / max_amount + 1;
+        let elems_in_node = (keys.len() / nodes_amount) as u64;
+        trace!(
+            "last node has {} less keys",
+            elems_in_node - elems_in_node % nodes_amount as u64
+        );
+
+        let leaf_nodes_compressed = keys
+            .chunks(elems_in_node as usize)
+            .enumerate()
+            .map(|(i, keys)| {
+                let cur_offset = leaves_offset + elems_in_node * leaf_size * i as u64;
+                (keys[0].as_ref().to_vec(), cur_offset)
+            })
+            .collect::<Vec<(Vec<u8>, u64)>>();
+
+        let mut buf = Vec::new();
+        let root_offset = Self::build_tree(leaf_nodes_compressed, tree_offset, &mut buf)?;
+        Ok((root_offset, buf))
+    }
+
+    pub(super) fn build_tree(
+        nodes_arr: Vec<(Vec<u8>, u64)>,
+        tree_offset: u64,
+        buf: &mut Vec<u8>,
+    ) -> Result<u64> {
+        if nodes_arr.len() == 1 {
+            return Ok(Self::prep_root(nodes_arr[0].1, tree_offset, buf));
+        }
+        let max_amount = Self::max_nonleaf_node_capacity(nodes_arr[0].0.len());
+        let nodes_amount = (nodes_arr.len() - 1) / max_amount + 1;
+        let elems_in_node = nodes_arr.len() / nodes_amount;
+        let new_nodes = nodes_arr
+            .chunks(elems_in_node)
+            .map(|keys| {
+                let offset = tree_offset + buf.len() as u64;
+                let min_key = keys[0].0.clone();
+                let offsets = keys.iter().map(|(_, offset)| *offset).collect();
+                let keys = keys.iter().skip(1).map(|(key, _)| key.clone()).collect();
+                let node = Node::new(keys, offsets);
+                buf.extend_from_slice(&node.serialize().expect("failed to serialize node"));
+                (min_key, offset)
+            })
+            .collect();
+        Self::build_tree(new_nodes, tree_offset, buf)
+    }
+
+    fn max_leaf_node_capacity(key_size: usize) -> usize {
+        let offset_size = std::mem::size_of::<u64>();
+        BLOCK_SIZE / (key_size + offset_size) - 1
+    }
+
+    fn max_nonleaf_node_capacity(key_size: usize) -> usize {
+        let offset_size = std::mem::size_of::<u64>();
+        let meta_size = std::mem::size_of::<NodeMeta>();
```

Comment:
Are you sure `size_of` is identical to `bincode::serialized_size()`?

Reply:
In this case - yes, but you're right, that it's better to use serialized_size, changed that